### PR TITLE
[Refactor] simpilified Operator.Builder implement

### PR DESCRIPTION
**Pull request by Seaven**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

Refactor Operator.Builder, because I need modify 30+ child-class construct method when I want to add a member variables in LogicalOperator！！！ I need modify 10+ child-class construct method when I want add a member variables in LogicalScanOperator！！！

before:
```
class operator {
    //... Member variables
    A
    B
    C

    operator(builder) {
        super(....., ...All member variables)
    }

    class Builder {
        //... once again: same Member variables

        //... set function
        //... set function
        Builder setXXX(xxx) {
            this.xxx = xxx
        }

        operator builder() {
            return new operator(builder)
        }
    }
}
```

after:
```
class operator {
    //... Member variables
    A
    B
    C

    operator() {
        super(TYPE)
    }

    class Builder {
        protected operator builder = new operator();

        //... set function
        Builder setXXX(xxx) {
            builder.xxx = xxx
        }

        operator builder() {
            return builder;
            builder = null;
        }
    }
}
```

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 189 / 236 (80.08%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalTreeAnchorOperator.java|0|2|00.00%|[55, 59]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalAssertOneRowOperator.java|0|5|00.00%|[46, 104, 110, 111, 112]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalEsScanOperator.java|0|6|00.00%|[35, 53, 74, 80, 81, 82]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalHudiScanOperator.java|0|4|00.00%|[51, 82, 89, 90]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalIcebergScanOperator.java|0|3|00.00%|[48, 71, 78]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalFileScanOperator.java|0|3|00.00%|[49, 80, 87]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalLimitOperator.java|0|7|00.00%|[51, 52, 139, 145, 146, 151, 155]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalCTEProduceOperator.java|0|6|00.00%|[92, 96, 100, 101, 105, 106]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalDeltaLakeScanOperator.java|0|3|00.00%|[42, 65, 71]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/OperatorBuilderFactory.java|3|4|75.00%|[27]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/Operator.java|12|14|85.71%|[177, 181]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalScanOperator.java|8|9|88.89%|[190]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalAggregationOperator.java|23|25|92.00%|[305, 309]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalTopNOperator.java|15|16|93.75%|[218]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|14|15|93.33%|[198]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalSchemaScanOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalJoinOperator.java|13|13|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalValuesOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalMysqlScanOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalExceptOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalUnionOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalWindowOperator.java|18|18|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalCTEConsumeOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalHiveScanOperator.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalFilterOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalIntersectOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalMetaScanOperator.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalCTEAnchorOperator.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/stream/LogicalBinlogScanOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalSetOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalRepeatOperator.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalProjectOperator.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalTableFunctionOperator.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalJDBCScanOperator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneScanColumnRule.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalApplyOperator.java|18|18|100.00%|[]|

**Review by packy92 — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/OperatorBuilderFactory.java [hunk 1]
```diff
@@ -16,94 +16,15 @@
...
-            return new LogicalProjectOperator.Builder();
-        } else if (operator instanceof LogicalBinlogScanOperator) {
-            return new LogicalBinlogScanOperator.Builder();
-        } else {
+    public static <T extends Operator.Builder> T build(Operator operator) {
+        String clazzName = operator.getClass().getName();
+        String builderName = clazzName + "$Builder";
+
+        try {
+            return (T) Class.forName(builderName).getConstructor().newInstance();
```
Comment (liuyehcf) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/OperatorBuilderFactory.java:26]: Does this has performance problems?

**Comment by liuyehcf** [comment]
I love this refactor. And I've noticed that you've put final to each operator's field, but some fields are of Collection type, which can be modified. Are you gonna to step further for replacing all the collections to it's immutable versions?

**Review by liuyehcf — APPROVED** [review]

**Comment by Seaven** [comment]
> I love this refactor. And I've noticed that you've put key word `final` to each operator's field, but some fields are of collection type, which can still be modified. Are you gonna to step further for replacing all the collections to it's immutable versions?

emm, we can modify it later and step by step, i think will be easy than before

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/OperatorBuilderFactory.java (hunk 1 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/OperatorBuilderFactory.java:26]: yeah, will slow X10 than if-else, but the cost is still low, about 100ms/10w call, I think the performance is OK, and you?

Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/OperatorBuilderFactory.java:26]: BTW, we can cache all constructor by map in JVM startup

Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/OperatorBuilderFactory.java:26]: add a cache is better

**Review by stdpain — APPROVED** [review]

**Pull request merged**